algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,342 @@
+    else {
+      val input = buckets.dropWhile(_.size == 0)
+      val bucketSize +: tail = desired
+      val remaining = drop(bucketSize, input)
+      input.head.copy(size = bucketSize) +: rebucket(remaining, tail)
+    }
+
+  /**
+   * @param toDrop total count to remove from the left of `input`.
+   * @param input buckets
+   * @return Vector with buckets, or pieces of buckets, with sizes
+   *         totalling `toDrop` items removed from the head. If an
+   *         element wasn't fully consumed, the remainder will be
+   *         stuck back onto the head.
+   */
+  @tailrec private[this] def drop(toDrop: Long, input: Vector[Bucket]): Vector[Bucket] = {
+    val (b @ Bucket(count, _)) +: tail = input
+    (toDrop - count) match {
+      case 0 => tail
+      case x if x < 0 => b.copy(size = -x) +: tail
+      case x if x > 0 => drop(x, tail)
+    }
+  }
+}
+
+/**
+ * @param conf the config values for this isntance.
+ * @param buckets Vector of timestamps of each (powers of 2)
+ *                ticks. This is the core of the exponential
+ *                histogram representation. See [[Canonical]] for
+ *                more info.
+ * @param total total ticks tracked, == buckets.map(_.size).sum
+ * @param time current timestamp of this instance.
+ *             Used with conf.windowSize to expire buckets.
+ */
+case class ExpHist(conf: ExpHist.Config, buckets: Vector[ExpHist.Bucket], total: Long, time: Long) {
+  import ExpHist.Bucket
+
+  /**
+   * Steps this instance forward to the new supplied time. Any
+   * buckets with a timestamp <= (newTime - conf.windowSize) will be
+   * evicted.
+   *
+   * @param newTime the new current time.
+   * @return ExpHist instance stepped forward to newTime.
+   */
+  def step(newTime: Long): ExpHist =
+    if (newTime <= time) this
+    else {
+      val (dropped, filtered) = conf.dropExpired(buckets, newTime)
+      copy(time = newTime, buckets = filtered, total = total - dropped)
+    }
+
+  /**
+   * Increment ExpHist by 1 at the supplied timestamp.
+   */
+  def inc(timestamp: Long): ExpHist = add(1L, timestamp)
+
+  /**
+   * Increment ExpHist by delta at the supplied timestamp.
+   */
+  def add(delta: Long, timestamp: Long): ExpHist = {
+    val self = step(timestamp)
+    if (delta == 0) self
+    else self.addAllWithoutStep(Vector(Bucket(delta, timestamp)), delta)
+  }
+
+  /**
+   * Efficiently add many buckets at once.
+   *
+   * @param unsorted [bucket]. All timestamps must be >= this.time.
+   * @return ExpHist instance with all buckets added, stepped
+   *         forward to the most timestamp in `unsorted`.
+   */
+  def addAll(unsorted: Vector[Bucket]): ExpHist =
+    if (unsorted.isEmpty) this
+    else {
+      val sorted = unsorted.sorted(Ordering[Bucket].reverse)
+      val delta = sorted.map(_.size).sum
+      val timestamp = sorted.head.timestamp
+      if (delta == 0)
+        step(timestamp)
+      else {
+        addAllWithoutStep(sorted, delta).step(timestamp)
```

Comment:
this was a little subtle. We have to step AFTER adding because the range of timestamps of the new items we're pushing in might be larger than windowSize.